File: src/TypeScript/src/numerics/wide-integer.ts
export type ByteOrder = "little-endian" | "big-endian";

function modulus(width: number): bigint { return 1n << BigInt(width); }
function normalizeBits(value: bigint, width: number): bigint { const mod = modulus(width); const result = value % mod; return result < 0n ? result + mod : result; }

/** Shared method surface for fixed-width integer values. */
export abstract class FixedWidthInteger<TSelf extends FixedWidthInteger<TSelf>> {
    readonly #bits: bigint;
    public abstract readonly width: number;
    public abstract readonly signed: boolean;
    protected constructor(bits: bigint) { this.#bits = bits; }
    protected abstract create(value: bigint): TSelf;
    protected normalizedBits(): bigint { return normalizeBits(this.#bits, this.width); }
    public toBigInt(): bigint {
        const bits = this.normalizedBits();
        if (!this.signed) return bits;
        const sign = 1n << BigInt(this.width - 1);
        return (bits & sign) === 0n ? bits : bits - modulus(this.width);
    }
    public toUnsignedBigInt(): bigint { return this.normalizedBits(); }
    public equals(other: TSelf): boolean { return this.normalizedBits() === other.normalizedBits(); }
    public get isZero(): boolean { return this.normalizedBits() === 0n; }
    public get isNegative(): boolean { return this.signed && this.toBigInt() < 0n; }
    public get isEven(): boolean { return (this.normalizedBits() & 1n) === 0n; }
    public get isOdd(): boolean { return !this.isEven; }
    public get isPowerOfTwo(): boolean { const value = this.toBigInt(); return value > 0n && (value & (value - 1n)) === 0n; }
    public compareTo(other: TSelf): number { const left = this.toBigInt(); const right = other.toBigInt(); return left < right ? -1 : left > right ? 1 : 0; }
    public add(other: TSelf): TSelf { return this.create(this.toBigInt() + other.toBigInt()); }
    public subtract(other: TSelf): TSelf { return this.create(this.toBigInt() - other.toBigInt()); }
    public multiply(other: TSelf): TSelf { return this.create(this.toBigInt() * other.toBigInt()); }
    public divide(other: TSelf): TSelf { if (other.toBigInt() === 0n) throw new RangeError("Division by zero."); if (this.signed && this.toBigInt() === -(1n << BigInt(this.width - 1)) && other.toBigInt() === -1n) throw new RangeError("Signed division overflow."); return this.create(this.toBigInt() / other.toBigInt()); }
    public remainder(other: TSelf): TSelf { if (other.toBigInt() === 0n) throw new RangeError("Division by zero."); if (this.signed && this.toBigInt() === -(1n << BigInt(this.width - 1)) && other.toBigInt() === -1n) throw new RangeError("Signed remainder overflow."); return this.create(this.toBigInt() % other.toBigInt()); }
    public divRem(other: TSelf): { readonly quotient: TSelf; readonly remainder: TSelf } { return { quotient: this.divide(other), remainder: this.remainder(other) }; }
    public negate(): TSelf { return this.create(-this.toBigInt()); }
    public increment(): TSelf { return this.create(this.toBigInt() + 1n); }
    public decrement(): TSelf { return this.create(this.toBigInt() - 1n); }
    public abs(): TSelf { const value = this.toBigInt(); if (this.signed && value === -(1n << BigInt(this.width - 1))) throw new RangeError("Absolute value overflow."); return this.create(value < 0n ? -value : value); }
    public bitwiseAnd(other: TSelf): TSelf { return this.create(this.normalizedBits() & other.normalizedBits()); }
    public bitwiseOr(other: TSelf): TSelf { return this.create(this.normalizedBits() | other.normalizedBits()); }
    public bitwiseXor(other: TSelf): TSelf { return this.create(this.normalizedBits() ^ other.normalizedBits()); }
    public bitwiseNot(): TSelf { return this.create(~this.normalizedBits()); }
    public shiftLeft(count: number): TSelf { const shift = BigInt(((count % this.width) + this.width) % this.width); return this.create(this.normalizedBits() << shift); }
    public shiftRight(count: number): TSelf { const shift = BigInt(((count % this.width) + this.width) % this.width); return this.create(this.signed ? this.toBigInt() >> shift : this.normalizedBits() >> shift); }
    public rotateLeft(count: number): TSelf { const normalized = ((count % this.width) + this.width) % this.width; if (normalized === 0) return this.create(this.normalizedBits()); const shift = BigInt(normalized); return this.create((this.normalizedBits() << shift) | (this.normalizedBits() >> BigInt(this.width - normalized))); }
    public rotateRight(count: number): TSelf { return this.rotateLeft(-count); }
    public checkedAdd(other: TSelf): TSelf { return this.checked(this.toBigInt() + other.toBigInt()); }
    public checkedSubtract(other: TSelf): TSelf { return this.checked(this.toBigInt() - other.toBigInt()); }
    public checkedMultiply(other: TSelf): TSelf { return this.checked(this.toBigInt() * other.toBigInt()); }
    public checkedNegate(): TSelf { return this.checked(-this.toBigInt()); }
    public checked(value: bigint): TSelf { const minimum = this.signed ? -(1n << BigInt(this.width - 1)) : 0n; const maximum = this.signed ? (1n << BigInt(this.width - 1)) - 1n : modulus(this.width) - 1n; if (value < minimum || value > maximum) throw new RangeError(`Value does not fit in ${this.width} bits.`); return this.create(value); }
    public getShortestBitLength(): number { const value = this.toBigInt(); if (value === 0n) return 0; if (!this.signed) return value.toString(2).length; if (value > 0n) return value.toString(2).length + 1; return (~value).toString(2).length + 1; }
    public leadingZeroCount(): number { const bits = this.normalizedBits(); return bits === 0n ? this.width : this.width - bits.toString(2).length; }
    public trailingZeroCount(): number { let bits = this.normalizedBits(); if (bits === 0n) return this.width; let count = 0; while ((bits & 1n) === 0n) { bits >>= 1n; count++; } return count; }
    public popCount(): number { let bits = this.normalizedBits(), count = 0; while (bits !== 0n) { bits &= bits - 1n; count++; } return count; }
    public log2(): number { const value = this.toBigInt(); if (value < 0n) throw new RangeError("Log2 is undefined for negative values."); return value === 0n ? 0 : value.toString(2).length - 1; }
    public getByteCount(): number { return this.width / 8; }
    public toBytes(order: ByteOrder = "little-endian"): Uint8Array {
        const bytes = new Uint8Array(this.getByteCount()); let value = this.normalizedBits();
        for (let index = 0; index < bytes.length; index++) { const target = order === "little-endian" ? index : bytes.length - index - 1; bytes[target] = Number(value & 0xffn); value >>= 8n; }
        return bytes;
    }
    public toString(radix = 10): string { if (radix < 2 || radix > 36) throw new RangeError("Radix must be between 2 and 36."); return (radix === 16 ? this.normalizedBits() : this.toBigInt()).toString(radix); }
    public format(format = "G"): string {
        const match = /^([gGdDnNxX])([0-9]*)$/.exec(format); if (match === null) throw new RangeError("Unsupported integer format.");
        const code = match[1]!; const precision = match[2] === "" ? (code === "N" || code === "n" ? 2 : 0) : Number(match[2]);
        if ((code === "G" || code === "g") && precision !== 0) throw new RangeError("General integer format does not accept precision.");
        if (code === "X" || code === "x") { const text = this.normalizedBits().toString(16).padStart(precision, "0"); return code === "X" ? text.toUpperCase() : text; }
        const negative = this.toBigInt() < 0n; const digits = (negative ? -this.toBigInt() : this.toBigInt()).toString().padStart(precision, "0");
        if (code === "N" || code === "n") { const grouped = digits.replace(/\B(?=(\d{3})+(?!\d))/g, ","); return `${negative ? "-" : ""}${grouped}${precision === 0 ? "" : `.${"0".repeat(precision)}`}`; }
        return `${negative ? "-" : ""}${digits}`;
    }
}

function parseInteger(text: string, radix: number): bigint {
    const trimmed = text.trim(); if (trimmed.length === 0) throw new SyntaxError("The input is not an integer.");
    if (radix === 10) { if (!/^[+-]?\d+$/.test(trimmed)) throw new SyntaxError("The input is not a decimal integer."); return BigInt(trimmed); }
    const negative = trimmed.startsWith("-"); const unsigned = /^[+-]/.test(trimmed) ? trimmed.slice(1) : trimmed;
    const digits = radix === 16 ? /^[0-9a-f]+$/i : radix === 2 ? /^[01]+$/ : undefined;
    if (digits === undefined || !digits.test(unsigned)) throw new SyntaxError("The input is not valid for the requested radix.");
    const value = BigInt(`${radix === 16 ? "0x" : "0b"}${unsigned}`); return negative ? -value : value;
}

function parseFixed(text: string, radix: number, width: number, signed: boolean): bigint {
    const value = parseInteger(text, radix);
    const explicitNegative = text.trim().startsWith("-");
    const minimum = signed ? -(1n << BigInt(width - 1)) : 0n;
    const maximum = signed && (radix === 10 || explicitNegative) ? (1n << BigInt(width - 1)) - 1n : (1n << BigInt(width)) - 1n;
    if (value < minimum || value > maximum) throw new RangeError(`Value does not fit in ${width} bits.`);
    return value;
}

function fromBytes(bytes: Uint8Array, width: number, signed: boolean, order: ByteOrder): bigint {
    if (bytes.byteLength !== width / 8) throw new RangeError(`Expected exactly ${width / 8} bytes.`);
    let bits = 0n; for (let index = 0; index < bytes.length; index++) { const source = order === "big-endian" ? index : bytes.length - index - 1; bits = (bits << 8n) | BigInt(bytes[source]!); }
    if (signed && (bits & (1n << BigInt(width - 1))) !== 0n) return bits - modulus(width); return bits;
}

abstract class ConcreteWide<TSelf extends ConcreteWide<TSelf>> extends FixedWidthInteger<TSelf> {
    protected constructor(value: bigint | number | string) { super(BigInt(value)); }
}

export class UInt256 extends ConcreteWide<UInt256> { public readonly width = 256; public readonly signed = false; public constructor(value: bigint | number | string = 0n) { super(value); } protected create(value: bigint): UInt256 { return new UInt256(value); } public static parse(text: string, radix = 10): UInt256 { return new UInt256(parseFixed(text, radix, 256, false)); } public static fromBytes(bytes: Uint8Array, order: ByteOrder = "little-endian"): UInt256 { return new UInt256(fromBytes(bytes, 256, false, order)); } public static get minValue(): UInt256 { return new UInt256(0n); } public static get maxValue(): UInt256 { return new UInt256((1n << 256n) - 1n); } }
export class Int256 extends ConcreteWide<Int256> { public readonly width = 256; public readonly signed = true; public constructor(value: bigint | number | string = 0n) { super(value); } protected create(value: bigint): Int256 { return new Int256(value); } public static parse(text: string, radix = 10): Int256 { return new Int256(parseFixed(text, radix, 256, true)); } public static fromBytes(bytes: Uint8Array, order: ByteOrder = "little-endian"): Int256 { return new Int256(fromBytes(bytes, 256, true, order)); } public static get minValue(): Int256 { return new Int256(-(1n << 255n)); } public static get maxValue(): Int256 { return new Int256((1n << 255n) - 1n); } }
export class UInt512 extends ConcreteWide<UInt512> { public readonly width = 512; public readonly signed = false; public constructor(value: bigint | number | string = 0n) { super(value); } protected create(value: bigint): UInt512 { return new UInt512(value); } public static parse(text: string, radix = 10): UInt512 { return new UInt512(parseFixed(text, radix, 512, false)); } public static fromBytes(bytes: Uint8Array, order: ByteOrder = "little-endian"): UInt512 { return new UInt512(fromBytes(bytes, 512, false, order)); } public static get minValue(): UInt512 { return new UInt512(0n); } public static get maxValue(): UInt512 { return new UInt512((1n << 512n) - 1n); } }
export class Int512 extends ConcreteWide<Int512> { public readonly width = 512; public readonly signed = true; public constructor(value: bigint | number | string = 0n) { super(value); } protected create(value: bigint): Int512 { return new Int512(value); } public static parse(text: string, radix = 10): Int512 { return new Int512(parseFixed(text, radix, 512, true)); } public static fromBytes(bytes: Uint8Array, order: ByteOrder = "little-endian"): Int512 { return new Int512(fromBytes(bytes, 512, true, order)); } public static get minValue(): Int512 { return new Int512(-(1n << 511n)); } public static get maxValue(): Int512 { return new Int512((1n << 511n) - 1n); } }
export class UInt1024 extends ConcreteWide<UInt1024> { public readonly width = 1024; public readonly signed = false; public constructor(value: bigint | number | string = 0n) { super(value); } protected create(value: bigint): UInt1024 { return new UInt1024(value); } public static parse(text: string, radix = 10): UInt1024 { return new UInt1024(parseFixed(text, radix, 1024, false)); } public static fromBytes(bytes: Uint8Array, order: ByteOrder = "little-endian"): UInt1024 { return new UInt1024(fromBytes(bytes, 1024, false, order)); } public static get minValue(): UInt1024 { return new UInt1024(0n); } public static get maxValue(): UInt1024 { return new UInt1024((1n << 1024n) - 1n); } }
export class Int1024 extends ConcreteWide<Int1024> { public readonly width = 1024; public readonly signed = true; public constructor(value: bigint | number | string = 0n) { super(value); } protected create(value: bigint): Int1024 { return new Int1024(value); } public static parse(text: string, radix = 10): Int1024 { return new Int1024(parseFixed(text, radix, 1024, true)); } public static fromBytes(bytes: Uint8Array, order: ByteOrder = "little-endian"): Int1024 { return new Int1024(fromBytes(bytes, 1024, true, order)); } public static get minValue(): Int1024 { return new Int1024(-(1n << 1023n)); } public static get maxValue(): Int1024 { return new Int1024((1n << 1023n) - 1n); } }

export type WideInteger = UInt256 | Int256 | UInt512 | Int512 | UInt1024 | Int1024;

/** Fixed-width byte conversion helpers. */
export const BitConverterEx: {
    getBytes(value: WideInteger, order?: ByteOrder): Uint8Array;
    tryWriteBytes(value: WideInteger, destination: Uint8Array, order?: ByteOrder): boolean;
} = {
    getBytes(value: WideInteger, order: ByteOrder = "little-endian"): Uint8Array { return value.toBytes(order); },
    tryWriteBytes(value: WideInteger, destination: Uint8Array, order: ByteOrder = "little-endian"): boolean { const bytes = value.toBytes(order); if (destination.byteLength < bytes.byteLength) return false; destination.set(bytes); return true; },
};

File: src/OCaml/lib/numerics/wide_integer.ml
type byte_order = Little_endian | Big_endian

module type S = sig
  type t

  val width : int
  val signed : bool
  val zero : t
  val one : t
  val min_value : t
  val max_value : t
  val of_int : int -> t
  val of_z : Z.t -> t
  val of_z_checked : Z.t -> t
  val to_z : t -> Z.t
  val to_unsigned_z : t -> Z.t
  val parse : ?radix:int -> string -> t
  val to_string : ?radix:int -> t -> string
  val equal : t -> t -> bool
  val compare : t -> t -> int
  val add : t -> t -> t
  val sub : t -> t -> t
  val mul : t -> t -> t
  val div : t -> t -> t
  val rem : t -> t -> t
  val div_rem : t -> t -> t * t
  val negate : t -> t
  val abs : t -> t
  val succ : t -> t
  val pred : t -> t
  val checked_add : t -> t -> t
  val checked_sub : t -> t -> t
  val checked_mul : t -> t -> t
  val checked_negate : t -> t
  val logand : t -> t -> t
  val logor : t -> t -> t
  val logxor : t -> t -> t
  val lognot : t -> t
  val shift_left : t -> int -> t
  val shift_right : t -> int -> t
  val rotate_left : t -> int -> t
  val rotate_right : t -> int -> t
  val is_zero : t -> bool
  val is_negative : t -> bool
  val is_even : t -> bool
  val is_power_of_two : t -> bool
  val shortest_bit_length : t -> int
  val leading_zero_count : t -> int
  val trailing_zero_count : t -> int
  val pop_count : t -> int
  val log2 : t -> int
  val byte_count : int
  val to_bytes : ?byte_order:byte_order -> t -> bytes
  val of_bytes : ?byte_order:byte_order -> bytes -> t
end

module type Parameters = sig
  val width : int
  val signed : bool
end

module Make (Parameters : Parameters) : S = struct
  type t = Z.t

  let width = Parameters.width
  let signed = Parameters.signed
  let modulus = Z.shift_left Z.one width
  let sign_bit = Z.shift_left Z.one (width - 1)
  let minimum = if signed then Z.neg sign_bit else Z.zero
  let maximum = if signed then Z.pred sign_bit else Z.pred modulus
  let byte_count = width / 8

  let unsigned value =
    let reduced = Z.erem value modulus in
    if Z.sign reduced < 0 then Z.add reduced modulus else reduced

  let normalize value =
    let bits = unsigned value in
    if signed && Z.compare bits sign_bit >= 0 then Z.sub bits modulus else bits

  let of_z = normalize

  let of_z_checked value =
    if Z.compare value minimum < 0 || Z.compare value maximum > 0 then
      invalid_arg (Printf.sprintf "value does not fit in %d bits" width);
    value

  let to_z value = value
  let to_unsigned_z = unsigned
  let of_int value = of_z (Z.of_int value)
  let zero = of_int 0
  let one = of_int 1
  let min_value = minimum
  let max_value = maximum
  let equal = Z.equal
  let compare = Z.compare
  let add left right = of_z (Z.add left right)
  let sub left right = of_z (Z.sub left right)
  let mul left right = of_z (Z.mul left right)

  let check_division dividend divisor =
    if Z.equal divisor Z.zero then raise Division_by_zero;
    if signed && Z.equal divisor Z.minus_one && Z.equal dividend minimum then
      invalid_arg "signed division overflow"

  let div left right =
    check_division left right;
    of_z (Z.div left right)

  let rem left right =
    check_division left right;
    of_z (Z.rem left right)

  let div_rem left right = (div left right, rem left right)
  let negate value = of_z (Z.neg value)

  let abs value =
    if signed && Z.equal value minimum then invalid_arg "absolute value overflow";
    of_z (Z.abs value)

  let succ value = add value one
  let pred value = sub value one
  let checked_add left right = of_z_checked (Z.add left right)
  let checked_sub left right = of_z_checked (Z.sub left right)
  let checked_mul left right = of_z_checked (Z.mul left right)
  let checked_negate value = of_z_checked (Z.neg value)
  let logand left right = of_z (Z.logand (unsigned left) (unsigned right))
  let logor left right = of_z (Z.logor (unsigned left) (unsigned right))
  let logxor left right = of_z (Z.logxor (unsigned left) (unsigned right))
  let lognot value = of_z (Z.lognot (unsigned value))

  let normalized_shift count =
    let reduced = count mod width in
    if reduced < 0 then reduced + width else reduced

  let shift_left value count = of_z (Z.shift_left (unsigned value) (normalized_shift count))

  let shift_right value count =
    let source = if signed then value else unsigned value in
    of_z (Z.shift_right source (normalized_shift count))

  let rotate_left value count =
    let shift = normalized_shift count in
    if shift = 0 then value
    else
      of_z
        (Z.logor
           (Z.shift_left (unsigned value) shift)
           (Z.shift_right (unsigned value) (width - shift)))

  let rotate_right value count = rotate_left value (-count)
  let is_zero value = Z.equal value Z.zero
  let is_negative value = signed && Z.sign value < 0
  let is_even value = not (Z.testbit (unsigned value) 0)
  let is_power_of_two value = Z.sign value > 0 && Z.equal (Z.logand value (Z.pred value)) Z.zero

  let shortest_bit_length value =
    if Z.equal value Z.zero then 0
    else if not signed then Z.numbits value
    else if Z.sign value > 0 then Z.numbits value + 1
    else Z.numbits (Z.lognot value) + 1

  let leading_zero_count value = width - Z.numbits (unsigned value)

  let trailing_zero_count value =
    let bits = unsigned value in
    if Z.equal bits Z.zero then width else Z.trailing_zeros bits

  let pop_count value = Z.popcount (unsigned value)

  let log2 value =
    if Z.sign value < 0 then invalid_arg "log2 is undefined for negative values";
    if Z.equal value Z.zero then 0 else Z.numbits value - 1

  let to_bytes ?(byte_order = Little_endian) value =
    let result = Bytes.make byte_count '\000' in
    let bits = unsigned value in
    for index = 0 to byte_count - 1 do
      let byte = Z.to_int (Z.extract bits (index * 8) 8) in
      let destination =
        match byte_order with Little_endian -> index | Big_endian -> byte_count - index - 1
      in
      Bytes.set result destination (Char.chr byte)
    done;
    result

  let of_bytes ?(byte_order = Little_endian) bytes =
    if Bytes.length bytes <> byte_count then
      invalid_arg (Printf.sprintf "expected exactly %d bytes" byte_count);
    let result = ref Z.zero in
    for index = 0 to byte_count - 1 do
      let source =
        match byte_order with Little_endian -> index | Big_endian -> byte_count - index - 1
      in
      let byte = Z.of_int (Char.code (Bytes.get bytes source)) in
      result := Z.logor !result (Z.shift_left byte (index * 8))
    done;
    of_z !result

  let parse ?(radix = 10) text =
    if radix < 2 || radix > 36 then invalid_arg "radix must be between 2 and 36";
    let stripped = String.trim text in
    if String.length stripped = 0 then invalid_arg "input is not an integer";
    let value = Z.of_string_base radix stripped in
    let explicitly_negative = stripped.[0] = '-' in
    if signed && radix <> 10 && not explicitly_negative then (
      if Z.sign value < 0 || Z.compare value modulus >= 0 then
        invalid_arg (Printf.sprintf "value does not fit in %d bits" width);
      of_z value)
    else of_z_checked value

  let digits radix value =
    if Z.equal value Z.zero then "0"
    else
      let alphabet = "0123456789abcdefghijklmnopqrstuvwxyz" in
      let negative = Z.sign value < 0 in
      let remaining = ref (Z.abs value) in
      let buffer = Buffer.create 32 in
      let radix_z = Z.of_int radix in
      while not (Z.equal !remaining Z.zero) do
        let quotient, remainder = Z.ediv_rem !remaining radix_z in
        Buffer.add_char buffer alphabet.[Z.to_int remainder];
        remaining := quotient
      done;
      if negative then Buffer.add_char buffer '-';
      let reversed = Buffer.contents buffer in
      String.init (String.length reversed) (fun index ->
          reversed.[String.length reversed - index - 1])

  let to_string ?(radix = 10) value =
    if radix < 2 || radix > 36 then invalid_arg "radix must be between 2 and 36";
    digits radix (if signed && radix <> 10 then unsigned value else value)
end

module UInt256 = Make (struct
  let width = 256
  let signed = false
end)

module Int256 = Make (struct
  let width = 256
  let signed = true
end)

module UInt512 = Make (struct
  let width = 512
  let signed = false
end)

module Int512 = Make (struct
  let width = 512
  let signed = true
end)

module UInt1024 = Make (struct
  let width = 1024
  let signed = false
end)

module Int1024 = Make (struct
  let width = 1024
  let signed = true
end)

module Bit_converter_ex = struct
  let try_write_bytes (type value) (module Integer : S with type t = value)
      ?(byte_order = Little_endian) value destination ~offset =
    if offset < 0 || offset > Bytes.length destination - Integer.byte_count then false
    else (
      Bytes.blit (Integer.to_bytes ~byte_order value) 0 destination offset Integer.byte_count;
      true)
end
